pkg/io/confio/sg.go

```diff
@@ -11,116 +11,242 @@ import (
 	"github.com/IBM/vpc-go-sdk/vpcv1"
 
 	configModel "github.com/np-guard/cloud-resource-collector/pkg/ibm/datamodel"
+	"github.com/np-guard/models/pkg/ipblock"
 
 	"github.com/np-guard/vpc-network-config-synthesis/pkg/ir"
 	"github.com/np-guard/vpc-network-config-synthesis/pkg/utils"
 )
 
-func makeSGRuleItem(rule *ir.SGRule, i int, sgRemoteRef *vpcv1.SecurityGroupRuleRemoteSecurityGroupReference) vpcv1.SecurityGroupRuleIntf {
+const ResourceTypeNif = "network_interface"
+const ResourceTypeEndpointGateway = "endpoint_gateway"
+
+func deleteTargetFromSG(model *configModel.ResourcesContainerModel, sgIndex, targetIndex int) {
+	part1 := model.SecurityGroupList[sgIndex].Targets[:targetIndex]
+	if targetIndex == len(model.SecurityGroupList[sgIndex].Targets)-1 {
+		model.SecurityGroupList[sgIndex].Targets = part1
+	} else {
+		part2 := model.SecurityGroupList[sgIndex].Targets[targetIndex+1:]
+		model.SecurityGroupList[sgIndex].Targets = append(part1, part2...) //nolint:gocritic // append result is assigned to the same slice
+	}
+}
+
+func findAndDeleteTargetFromSG(model *configModel.ResourcesContainerModel, sgIndex int, id *string) {
+	for i := range model.SecurityGroupList[sgIndex].Targets {
+		if target, ok := model.SecurityGroupList[sgIndex].Targets[i].(*vpcv1.SecurityGroupTargetReference); ok && *target.ID == *id {
+			deleteTargetFromSG(model, sgIndex, i)
+			break
+		}
+	}
+}
+
+func lookupOrCreate(nameToSGRemoteRef map[string]*vpcv1.SecurityGroupRuleRemoteSecurityGroupReference,
+	name string) *vpcv1.SecurityGroupRuleRemoteSecurityGroupReference {
+	if sgRemoteRef, ok := nameToSGRemoteRef[name]; ok {
+		return sgRemoteRef
+	}
+	ref := allocateRef()
+	sgRemoteRef := &vpcv1.SecurityGroupRuleRemoteSecurityGroupReference{
+		ID:   ref.ID,
+		CRN:  ref.CRN,
+		Href: ref.Href,
+		Name: utils.Ptr(name),
+	}
+	return sgRemoteRef
+}
+
+func sgRemote(nameToSGRemoteRef map[string]*vpcv1.SecurityGroupRuleRemoteSecurityGroupReference,
+	rule *ir.SGRule) vpcv1.SecurityGroupRuleRemoteIntf {
+	st := rule.Remote.String()
+	switch t := rule.Remote.(type) {
+	case *ipblock.IPBlock:
+		if ir.IsIPAddress(t) {
+			return &vpcv1.SecurityGroupRuleRemoteIP{
+				Address: &st,
+			}
+		}
+		return &vpcv1.SecurityGroupRuleRemoteCIDR{
+			CIDRBlock: &st,
+		}
+	case ir.SGName:
+		return lookupOrCreate(nameToSGRemoteRef, ir.ChangeScoping(st))
+	}
+	return nil
+}
+
+func makeSGRuleItem(nameToSGRemoteRef map[string]*vpcv1.SecurityGroupRuleRemoteSecurityGroupReference,
+	rule *ir.SGRule, i int) vpcv1.SecurityGroupRuleIntf {
 	iPVersion := utils.Ptr(ipv4Const)
 	direction := direction(rule.Direction)
+	cidrAll := ipblock.CidrAll
+	local := &vpcv1.SecurityGroupRuleLocal{
+		CIDRBlock: &cidrAll,
+	}
```

Comment:
To be more consistent (and avoid code dup), also declare and assign remote variable here.

Reply:
`local` should always be `0.0.0.0/0`. `sgRemote` returns the remote of the specific rule.

Reply:
I meant something like `remote := sgRemote(nameToSGRemoteRef, rule)`, then use `remote` in each case of the switch.